### ISSUE TITLE
No progress when publishing many QoS 2 messages.

### ISSUE THREAD

**Issue by jemc** [erlio/vernemq#39]
This is another issue encountered after #37 was resolved.

My test case publishes 500 QoS 2 messages via the `libmosquitto` client API and waits for the publish handshakes to complete before continuing.  This works fine when using a mosquitto server with `max_inflight_messages=20` on both the server and client side.  When I substitute the mosqitto server for a VerneMQ server (with `max_inflight_messages=20`), only the first 20 handshakes complete, and the rest don't make further progress.

**Comment by dergraf** [comment]
Thanks for reporting.

max_inflight_messages is buggy (not completely or correctly implemented) in the version you're using..
the current version of vmq_server should fix this (although we haven't tested if we get the exact same semantics as mosquitto).

Upgrading to the newest vmq_server is a bit of hack currently. If you want to try please do the following:

rm -Rf _build/default/lib/vmq_server
./rebar3 upgrade vmq_server
./rebar3 upgrade vmq_server (yep twice)
make rel

ensure to delete the data/meta folder (as we introduced a backward incompatible change for the subscriber/retain storage)

you're now bleeding edge, maybe this will work for you, if not we'll fix that for the next release.

**Comment by dergraf** [comment]
@jemc, what is the status of this issue? Are you experiencing the same in 0.11.0?

**Comment by jemc** [comment]
@dergraf - is it possible yet to upgrade and test the latest server by "normal" means, or do I still have to use the "hack" upgrade process described above?

**Comment by dergraf** [comment]
@jemc It's best to start from a fresh clone, then the hack wont be necessary anymore. Sorry for the inconveniences.

**Comment by jemc** [comment]
@dergraf - This appears to still be a problem when working from a fresh clone - the same test is timing out in the same way.

**Comment by dergraf** [comment]
can you share the test script you're using?

**Comment by jemc** [comment]
@dergraf - Sure, it's the test suite for a public Ruby library I maintain, so it's available for anyone to run.
Here's how:

1. Install Ruby - it should be a modern version (I currently use `2.2.2p95`, but `>= 2.0` should work).
2. Install Bundler - `gem install bundler`.
3. Clone the library - `git clone https://github.com/jemc/ruby-mosq.git && cd ruby-mosq`
4. Install runtime and development dependencies - `bundle install`
5. Run the test suite (also downloads and compiles the `libmosquitto` library on the first invocation) - `bundle exec rake`

For me, the tests hang on the specific test, then fail after a 30 second timeout, with output that looks like this:
```
Mosq::Client ..................F

  1) Mosq::Client when connected can subscribe to many topics and publish many messages
     Failure/Error: subject.publish_many(topics.zip(payloads), qos: 2, retain: false)
     Mosq::FFI::Error::Timeout:
       timed out - waiting for publish responses
     # ./lib/mosq/client.rb:421:in `fetch_responses'
     # ./lib/mosq/client.rb:249:in `publish_many'
     # ./spec/client_spec.rb:195:in `block (3 levels) in <top (required)>'

Finished in 31.06 seconds (files took 0.39432 seconds to load)
19 examples, 1 failure
```

If you just want to read the source for the test case that fails, you can find it [here](https://github.com/jemc/ruby-mosq/blob/bb91d7da50561376e97a8bb419bbaadfe92016b1/spec/client_spec.rb#L183-L215).

**Comment by dergraf** [comment]
Tried it with the test cases of your client library and found the problem. :)

Here is the thing. The ``max_inflight_messages`` in mosquitto  only cares about the delivered publishes and handshakes. In VerneMQ we use ``max_inflight_messages`` to also control incoming publishes... but this might be wrong as we can control the message rate using another feature as well.

**Comment by dergraf** [comment]
fixed in 0.12.0

**Comment by jemc** [comment]
Yep, the test cases now pass for me as well.  Thanks for your attention!
